### PR TITLE
Portals

### PR DESCRIPTION
**Pull request by blazej-smorawski**
Still WIP, please read the comments in the code. If you want some changes it would be best if you could commit them on your own :)

### CODE REVIEW

File: src/runtime.py [hunk 1]
```diff
@@ -1,31 +1,35 @@
-from src.objects.agent import Agent
+from src.actions.next_scene import NextScene
 from src.common.enums import Direction
```
Comment (Krystian030) [inline src/runtime.py:2]: unused import

Comment (Krystian030) [inline src/runtime.py]: nieużywane importy (w Pycharmie można użyć skrótu ctrl+alt+o do optymalizacji importów)

File: src/scene.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        self.objects_count = 0
+        self.objects_map = {}  # Point : [Object] Field
+        self.objects_dict = {}  # ID : Object
+        self.agent_locals = {}
+        self.runtime = runtime
+        self.__player: Agent = Agent(self)
+
+        # It's an example init sequence
+        # called each time the player goes through a portal
+        Portal(self, Point(random.randint(1,4), random.randint(1,4)))
```
Comment (Krystian030) [inline src/scene.py]: Te wszystkie wartości później trzeba będzie dać później do jakiegoś .env'a

File: src/objects/portal.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    '''
+
+    def __init__(self, scene, position: Point = Point(0, 0)) -> None:
+        super().__init__(position, scene)
+        self.walkable = True
+
+    def on_collision(self, other) -> None:
+        self.scene.runtime.next_scene.execute()
+        # Stop execution flow of `exec`
+        raise ExecInterrupt
```
Comment (Krystian030) [inline src/objects/portal.py]: proponuję, aby zmienić nazwę rzucanego wyjątku

File: src/actions/move.py [hunk 4]
```diff
@@ -1,13 +1,35 @@
...
-    def execute(self) -> str:
-        self.agent.properties.position += self.properties.direction.value
-        return super().execute()
+    def execute(self, scene) -> str:
+        new_position: Point = self.agent.properties.position + \
+            self.properties.direction.value
+
+        # Check if the destination is a `field` in the scene, and if it is -
+        # - it must be walkable
+        if scene[new_position] is not None and scene[new_position][0].walkable:
```
Comment (Krystian030) [inline src/actions/move.py]: magic number

**Review by Krystian030 — CHANGES_REQUESTED** [review]

File: src/actions/move.py (hunk 4 above)
Comment (blazej-smorawski) [reply src/actions/move.py]: Ta, było omawiane na spotkaniu w sobotę. Docelowo trzeba dodać abstrakcję `pola` na mapie, które zawiera inne rzeczy stojące na nim.

File: src/objects/portal.py (hunk 3 above)
Comment (blazej-smorawski) [reply src/objects/portal.py]: Na jaką?

File: src/scene.py (hunk 2 above)
Comment (blazej-smorawski) [reply src/scene.py]: Tymczasowa scena, zanim dodamy mapy z `.bmp`

Comment (Krystian030) [reply src/scene.py]: Ok

File: src/actions/move.py (hunk 4 above)
Comment (Krystian030) [reply src/actions/move.py]: Spoko, nie było mnie, to rozwiązuje

File: src/objects/portal.py (hunk 3 above)
Comment (Krystian030) [reply src/objects/portal.py]: Czekam na inwencje twórczą xd

File: src/objects/object.py [hunk 5]
```diff
@@ -3,13 +3,32 @@
...
         super().__init__(__class__.__name__)
-        self.properties.id: int = Object.__objects_count
+        self.properties.id: int = scene.objects_count
         self.properties.position: Point = position
+        self.scene = scene
+        self.walkable = False
 
-        Object.objects_dict[self.properties.id] = self
-        Object.__objects_count += 1
+        scene.add_object_to_map(self)
```
Comment (alpinus4) [inline src/objects/object.py:18]: i don't like passing scene to object constructor
it should be done outside object

File: src/runtime.py [hunk 6]
```diff
@@ -1,31 +1,35 @@
...
-    MAX_SPINS: int = 8
+    '''
+    Supplies basic execution of users code across various levels.
+    '''
 
     def __init__(self, agents_code: str) -> None:
         self.agents_code: str = agents_code
+        self.scene: Scene = None
+        self.scenes = []
+        self.next_scene = NextScene(self)
```
Comment (alpinus4) [inline src/runtime.py:15]: Also I don't like passing runtime to scene
If it requires passing self, it means it should be done in runtime itself

File: src/scene.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+
+class Scene():
+    '''
+    Object representing one level in the game.
+    It is responsible for storing all instances of `Object` class.
+    '''
+
+    def __init__(self, runtime) -> None:
+        self.objects_count = 0
```
Comment (alpinus4) [inline src/scene.py:16]: Let's create class `objects collection` aggregating all of `objects` related stuff
count, dict and map

File: src/actions/action.py [hunk 8]
```diff
@@ -1,16 +1,24 @@
...
+
+    def __init__(self) -> None:
         super().__init__(__class__.__name__)
-        self.properties.agent_id = agent_id
-        self.properties.target_id = target_id
-        self.agent: Agent = Object.objects_dict[agent_id]
-        self.target: Agent = Object.objects_dict[target_id]
 
-    def execute(self) -> str:
+    def execute(self, scene) -> None:
```
Comment (alpinus4) [inline src/actions/action.py:13]: Shouldn't there be `pass`?

File: src/runtime.py (hunk 1 above)
Comment (filip-szweda) [reply src/runtime.py]: Błażej is using VSCode, I think :p
Still, I think switching to Pycharm would be better for everyone involved.

File: usr/code.py [hunk 9]
```diff
@@ -0,0 +1,4 @@
+move(Direction.UP)
```
Comment (filip-szweda) [inline usr/code.py:1]: As previously, IMO this file should stay empty.
What do you think about loading user code from untracked, automatically created `.txt` file?

File: src/scene.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        self.objects_map = {}  # Point : [Object] Field
```
Comment (filip-szweda) [inline src/scene.py:17]: Instead of using comments, we can just declare class attributes before `__init__` along with their types.

File: src/objects/portal.py (hunk 3 above)
Comment (filip-szweda) [reply src/objects/portal.py]: Are we going to use it for anything other than scene change?
If not, maybe just let's name the exception `SceneChange`.

File: src/actions/move.py (hunk 4 above)
Comment (filip-szweda) [inline src/actions/move.py]: I think `scene` should have a method for checking if destination is plausible.
Somethin like `scene.is_field(new_pos)`.

File: src/actions/move.py [hunk 11]
```diff
@@ -1,13 +1,35 @@
...
+            scene[self.agent.properties.position].remove(self.agent)
```
Comment (filip-szweda) [inline src/actions/move.py]: `scene.move(object, old_pos, new_pos)`

File: src/actions/move.py [hunk 12]
```diff
@@ -1,13 +1,35 @@
...
+            scene[new_position].append(self.agent)
+            self.agent.properties.position = new_position
+
+            # Logging happens before collision detection
+            # because it may result in loading next scene,
+            # so we must ensure proper order of logs.
+            self.log()
+
+            for collision_object in scene[new_position]:
```
Comment (filip-szweda) [inline src/actions/move.py]: `scene.collide(object, pos)`

File: src/objects/agent.py [hunk 13]
```diff
@@ -1,10 +1,33 @@
...
+            'Direction': Direction,
+            'len': len,
+            'range': range,
+            'enumerate': enumerate,
+        }
+
+        # Sure, I know exec bad
+        # pylint: disable=exec-used
+        exec(
+            self.scene.runtime.agents_code,  # It violates lasagna code rule but I guess its ok
```
Comment (alpinus4) [inline src/objects/agent.py]: If it violates then let's don't do it
It's gonna become big, so making it messy now will make it really hard later
let's move `agents_code` to Agent
Agent shouldn't have the access to runtime

**Review by alpinus4 — CHANGES_REQUESTED** [review]

File: src/objects/object.py (hunk 5 above)
Comment (blazej-smorawski) [reply src/objects/object.py:18]: The `add_object_to_map` call could be easily moved outside `Object`. On the other hand the `Object` will still probably need referenece to scene. Other way would be to pass callbacks like `move` etc., but I rather have the reference.

File: src/runtime.py (hunk 6 above)
Comment (blazej-smorawski) [reply src/runtime.py:15]: But it's an `Action` for changing the state of the runtime, so it cannot operate without the reference.

File: src/actions/action.py (hunk 8 above)
Comment (blazej-smorawski) [reply src/actions/action.py:13]: A comment block is enough

File: usr/code.py (hunk 9 above)
Comment (blazej-smorawski) [reply usr/code.py:1]: I think that at the moment it doesn't matter.

File: src/scene.py (hunk 10 above)
Comment (blazej-smorawski) [reply src/scene.py:17]: I don't understand what you mean here, if you know how to do it better, please push the changes.

File: src/actions/move.py (hunk 4 above)
Comment (blazej-smorawski) [reply src/actions/move.py]: Sure, as I said - it is a subject to change, but I think that a `property` is a better way of describing an `Object` rather than creating come pointless function doing just `return object.walkable`.

File: src/objects/agent.py (hunk 13 above)
Comment (blazej-smorawski) [reply src/objects/agent.py]: Will see if it works, but I'm not a huge fan of passing an object from input to runtime to scene to agent.

File: src/scene.py (hunk 10 above)
Comment (filip-szweda) [reply src/scene.py:17]: Ok, will do.

File: usr/code.py (hunk 9 above)
Comment (filip-szweda) [reply usr/code.py:1]: Ok, but FYI I added https://gut-lynx.atlassian.net/browse/LNX-41 for the future.

File: src/runtime.py (hunk 1 above)
Comment (blazej-smorawski) [reply src/runtime.py]: Done

File: src/objects/portal.py (hunk 3 above)
Comment (blazej-smorawski) [reply src/objects/portal.py]: Chosen `MapExit`, because it will always quit a map but not always load next.

File: src/objects/object.py (hunk 5 above)
Comment (blazej-smorawski) [reply src/objects/object.py:18]: I won't change that now, if you still would like to see the changes we can make other task.

File: src/scene.py (hunk 7 above)
Comment (blazej-smorawski) [reply src/scene.py:16]: Sure, but not now :D [LNX-44](https://gut-lynx.atlassian.net/browse/LNX-44)

File: src/actions/move.py (hunk 12 above)
Comment (blazej-smorawski) [reply src/actions/move.py]: Done

File: src/actions/move.py (hunk 11 above)
Comment (blazej-smorawski) [reply src/actions/move.py]: Done

**Comment by blazej-smorawski** [comment]
TL;DR - Fixed, merge

Fixed most of the issues, but I think we should not add too many requests to a PR and rather focus on errors, etc. I created a ticket for some changes and we might for others, but we should merge it, so we won't be blocked.

**Review by Krystian030 — APPROVED** [review]
